packages/core/src/lib/metadata/index.ts

```diff
@@ -84,9 +89,19 @@ class Metadata extends BaseProtocol {
 
     const decodedResponse = this.decodeMetadataResponse(encodedResponse);
 
+    await this.savePeerShardInfo(peerId, decodedResponse);
+
     return decodedResponse;
   }
 
+  public async confirmOrAttemptHandshake(peerId: PeerId): Promise<void> {
+    if (this.checkHandshake(peerId)) return;
```

Comment:
you can simply replace it with `this.handshakesConfirmed.has(peerId)`

Reply:
not really as we can't directly compare two `peerId` -- usage of `peerId.equals(peerId2)` needs to be done